### PR TITLE
Material: Add "material.alphaHash" transparency mode

### PR DESCRIPTION
**Pull request by donmccurdy**
**Related issues:**
 
- https://github.com/mrdoob/three.js/issues/9977
- https://github.com/mrdoob/three.js/issues/19554

**Description:**

Proposal for adding `.alphaHash: boolean` and `.alphaHashScale: number` properties to all three.js Material types. This is an alternative method of transparency distinct from alpha blending (`.transparent=true`) and physically-based transmission (`.transmission > 0`), which is not a panacea but has some significant benefits:

1. Keeps objects in the 'opaque' render list
2. Avoids all sorting issues associated with alpha blending
3. Less overdraw than alpha blending
4. Less performance cost than transmission
5. Keeps objects visible through transmissive materials

The major drawback of the method is noise / grain, which can be reduced significantly with appropriate anti-aliasing. With or without AA, alpha hash is a good choice in several use cases:

1. Transitioning visibility of opaque objects without triggering sorting and backface problems during the transition. E.g. switching between LODs should not cause distracting flicker.
2. Hiding boundaries between two opaque objects, e.g. blades of grass fade into the ground, without requiring everything to be transparent.
3. Scenes with so much overlapping transparency that sorting issues are intractable

In these cases a bit of grain might be a reasonable tradeoff to avoid alpha blending issues — particularly if the transparent region is temporary, small, or distant from the camera.

The effect closely resembles Blender's *Blend Mode: Alpha Hashed* material property. Blender enables multi-sampling by default, which hides most of the grain, but this can be disabled in its renderer settings. The `.alphaHashScale: number` parameter should be used to tune the hash function to fit the multi-sampling method, with values in the range of 0.1–0.5 working well in my tests. When multi-sampling is not used, 1.0 is fine.

| # samples | three.js | blender |
|-----------|----------|---------|
| 1 | three_1 | blender_1 |
| 16 | three_16 | blender_16 |

**Questions:**

- [x] ~~Do we want to support `material.alphaHash`? We could also consider a NodeMaterial-only solution, as a separate node: `material.alphaTest = new AlphaHashNode( opacityNode, scale );`~~
- [x] ~~TAARenderPass is brightening the scene, and I'm not sure why. I've omitted all color management settings for now, to isolate that underlying problem. Any ideas here? Is this an appropriate type of multi-sampling to be using for this example? /cc @bhouston~~ See below.
- [x] ~~Better example? I used this [chloroplast scene](https://sketchfab.com/3d-models/chloroplast-42eec3ca7c074048b57f94ea9db7e7a8) from Sketchfab, mainly because it doesn't work well with alpha blending, for any choice of depthWrite.~~

**References:**

- https://casual-effects.com/research/Wyman2017Hashed/index.html

***

*This contribution is funded by [The New York Times](https://www.nytimes.com/).*

### CODE REVIEW

**Comment by bhouston** [comment]
This is really cool!  It is fast and the results are nice!

With regards to TAA causing things to get brighter when it is enabled, it appears to do this with or without your changes.  I checked here and it has that same bug in main: https://threejs.org/examples/?q=taa#webgl_postprocessing_taa. This wasn't the case a long time ago so some change in the last year or two must have broken it.  But it is unrelated to your changes.

**Comment by donmccurdy** [comment]
Thanks @bhouston! I'll file that as a separate issue then, I'm having some other trouble with color management in post-processing (unrelated to TAARenderPass) that may be worth looking at.

Assuming we'll find a fix so that the scene brightness doesn't fluctuate, I would be interested in everyone's feedback on whether we want to support this. Happy to clean up the example and implementation more if so. :)

**Comment by Mugen87** [comment]
> With regards to TAA causing things to get brighter when it is enabled,

@donmccurdy Can you please check if reverting #23671 fixes the issue? If so, `SSAARenderPass` might need a custom copy pass.

**Comment by donmccurdy** [comment]
Thanks @Mugen87! Unfortunately I don't see a difference after reverting #23671. The same thing happens in the [official TAA example](https://threejs.org/examples/?q=taa#webgl_postprocessing_taa) if I add this line:

```js
scene.background = new THREE.Color( 0x505050 );
```

**Comment by donmccurdy** [comment]
Possibly https://github.com/mrdoob/three.js/issues/9051?

File: examples/webgl_materials_alphahash.html [hunk 1]
```diff
@@ -0,0 +1,236 @@
...
+
+			import { OrbitControls } from './jsm/controls/OrbitControls.js';
+			import { GLTFLoader } from './jsm/loaders/GLTFLoader.js';
+			import { GUI } from './jsm/libs/lil-gui.module.min.js';
+
+			import { EffectComposer } from './jsm/postprocessing/EffectComposer.js';
+			import { RenderPass } from './jsm/postprocessing/RenderPass.js';
+			import { ShaderPass } from './jsm/postprocessing/ShaderPass.js';
+			import { TAARenderPass } from './jsm/postprocessing/TAARenderPass.js';
+			import { GammaCorrectionShader } from './jsm/shaders/GammaCorrectionShader.js';
```
Comment (LeviPesin) [inline examples/webgl_materials_alphahash.html:59]: `GammaCorrectionShader` is unused.

Comment (donmccurdy) [reply examples/webgl_materials_alphahash.html:59]: Lots to clean up here – I'll do cleanup if there's interest in adding the feature.

**Comment by donmccurdy** [comment]
Setting `scene.background` explicitly to black fixes the most noticeable flashing in the demo, although the model itself still shows a bit of flicker caused by the brightening issue I think.  Hopefully that is good enough to evaluate and consider the `.alphaHash` feature for now.

**Comment by hybridherbst** [comment]
Great to see! Typically transparency-through-noise can be improved with alpha to coverage, have you tried that? Could reduce the noise even more for low-ish AA values.

**Comment by donmccurdy** [comment]
Good point — to be honest I haven't really used alpha-to-coverage before. Assuming I've set that up right (alphaHash=true, alphaToCoverage=true, transparent=false, renderer.antialias=true) the difference is visible, but not in a way that seems obviously better or worse in this example:

| alphaToCoverage=true | alphaToCoverage=false |
|---|---|
| atc_on | atc_off |

Despite the checkbox in the screenshot, TAA is off for both. I'm not confident enough about setting up post-processing to combine TAA with WebGL2 anti-aliased render targets. So there's more grain than in the original screenshots — there are probably better ways to set this up and minimize noise.

**Comment by donmccurdy** [comment]
@gkjohnson from the paper, emphasis mine —

> User parameter `g_HashScale` controls the target noise scale in pixels (default 1.0).
Changing g_HashScale is useful if the chosen hash outputs noise at
another frequency. Also, _**when temporal antialiasing, using noise
below pixel scale (e.g., 0.3–0.5) allows for temporal averaging**_.

With TAA, I'm finding 0.1 looks pretty good. This may depend on the number of samples. Without TAA, I also see an improvement when decreasing the scale from 1 to 0.5, but not much change below that — unsure why that would happen.

The paper also discusses "Applications to Alpha-to-Coverage" but I haven't gotten into that area yet, it may require more work.

**Comment by donmccurdy** [comment]
Hmm, from one of the authors...

> To move to alpha to coverage, you need to independently compute the binary visibility tests for each sample and explicitly output them, e.g., via gl_SampleMask.  You cannot output an alpha value from the shader and let the hardware apply its default dithering.

https://twitter.com/_cwyman_/status/935743334244605953

I don't think WebGL 1.0 or 2.0 include `gl_SampleMask`, that's an OpenGL 4.0 feature. :/

**Comment by donmccurdy** [comment]
I've filed the issues about TAARenderPass at https://github.com/mrdoob/three.js/issues/24318 — it turns out we can work around them both by using a black scene background and keeping `pass.sampleLevel = 0`. Because `gl_SampleMask` is not available, alpha-to-coverage and multi-sampling can't reduce the noise associated with Hashed Alpha Testing, and so TAA seems to be the best option for reducing that noise. On higher-DPI displays or smaller surfaces, the noise is much less noticable and TAA may not be needed. 

Here's the preview link once again:

- [webgl / materials / alphahash](https://raw.githack.com/donmccurdy/three.js/feat-alphahash/examples/index.html?q=alpha#webgl_materials_alphahash)

Visually, this is comparable to Blender, so I probably won't make any further changes here unless we decide to move forward with the feature. If so, some further testing and cleanup would be needed.

**Comment by sdarpeng** [comment]
You should have this model to process your branch test.
@donmccurdy 
[hair_test1.zip](https://github.com/mrdoob/three.js/files/9972170/hair_test1.zip)

**Comment by donmccurdy** [comment]
I believe this is a useful improvement for the example shared by @sdarpeng as well. The appearance is stable (no sort issues) under different camera angles.

| before | after |
|---|---|
| Screenshot 2023-06-07 at 10 11 06 AM | Screenshot 2023-06-07 at 10 11 04 AM |

File: src/renderers/shaders/ShaderLib/meshphysical.glsl.js [hunk 2]
```diff
@@ -52,6 +58,12 @@ void main() {
 
 	vWorldPosition = worldPosition.xyz;
 
+#endif
+
+#ifdef USE_ALPHAHASH
+
+	vObjectPosition = transformed.xyz;
```
Comment (WestLangley) [inline src/renderers/shaders/ShaderLib/meshphysical.glsl.js:65]: Can the existing varying `vWorldPosition` be used, instead?

Also, what is the reason this is chunk is physical-material-specific?

Comment (donmccurdy) [reply src/renderers/shaders/ShaderLib/meshphysical.glsl.js:65]: Yes, and only for convenience of demonstrating the feature. I'll add it to other material types if we think this PR might move forward!

**Comment by Mugen87** [comment]
> Do we want to support material.alphaHash? We could also consider a NodeMaterial-only solution, as a separate node: material.alphaTest = new AlphaHashNode( opacityNode, scale );

I vote to add `Material.alphaHash`. In this way, we can provide at least one way in the core for addressing typical transparency artifacts related to alpha blending. 

IMO, TAA _should_ be used for acceptable image quality. But given the manageable amount of changes related to the core, I think it's worth an addition.

**Comment by mrdoob** [comment]
Yeah, this looks good to me 👍

As per the example, maybe something like [this one](https://threejs.org/examples/#webgl_instancing_raycast) but every sphere has a different color and alpha `0.5`?

File: src/renderers/shaders/ShaderLib/meshphysical.glsl.js (hunk 2 above)
Comment (donmccurdy) [reply src/renderers/shaders/ShaderLib/meshphysical.glsl.js:65]: @WestLangley Correction – `vWorldPosition` does not work as well, for reasons mentioned in the paper and apparent after I tested this:

> Hashing world-space coordinates provides stable noise for static geometry, and our early tests used world-space coordinates. However, this fails on dynamic geometry. Object-space coordinates give stable hashes for skinned and rigid transforms and dynamic cameras.

Given that we want the hash anchored to skinned and morphed surfaces, perhaps vObjectPosition is not the right name? We simply want to write the `position` attribute to a varying, would `vPosition` be appropriate?

**Comment by donmccurdy** [comment]
> As per the example, maybe something like [this one](https://threejs.org/examples/#webgl_instancing_raycast) but every sphere has a different color and alpha 0.5?

I might shuffle the order of instances to make the problems with alpha blending more obvious from any angle, but yes that sounds fine!

If there are no objections, I'd like to remove the `.alphaHashScale` property for now, using a hard-coded default instead. I haven't found any way to create a scene that benefits from changing its value, despite my earlier response, so including the property feels premature.

File: src/renderers/shaders/ShaderLib/meshphysical.glsl.js (hunk 2 above)
Comment (WestLangley) [reply src/renderers/shaders/ShaderLib/meshphysical.glsl.js:65]: Yes.

**Comment by donmccurdy** [comment]
To simplify things, I've removed the demo from this PR, and will create a new demo in a separate PR.

I've also moved the _alphaHash_ GLSL out of the _alphaTest_ shader chunks, and into its own chunks. Happy to revert that if we'd prefer otherwise.

**Comment by donmccurdy** [comment]
PR for the example:

- https://github.com/mrdoob/three.js/pull/26341

**Comment by sdarpeng** [comment]
Greaaaaat Job, guys.

**Comment by sdarpeng** [comment]
https://github.com/mrdoob/three.js/assets/55733077/f7c09e5c-7d67-4e85-9795-a6a74fbfe8d6

The new three.js alphaHash transparency looks fucking good.

**Pull request merged**